### DISCUSSION TITLE
Dynamic branching - I am at a loss

### DISCUSSION THREAD

**Discussion by rkrug** [ropensci/targets#1296] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I asked this question already on stack overflow (https://stackoverflow.com/questions/78582591/branching-i-am-at-a-loss) but only afterwards I was posted to here, which seems to be the best site to ask the question.

So - here is the question:

I need branching for my target pipeline (at least I think so)

My definition for the target pipeline looks as follows:

    list(
      # ILK
      tar_target(ilk_file, file.path(".", "input", "file_1.csv"), format = "file"),
      tar_target(ilk_bibliography, bibliography_metrics(ilk_file, mc.cores = 1)),
      tar_target(ilk_metrics, dois_measures(ilk_bibliography)),
      tar_target(plot_pub_year_ilk, plot_publication_year(ilk_bibliography)),
      tar_target(plot_oa_status_ilk, plot_oa_status(ilk_bibliography)),
      tar_target(plot_top_journals_data_ilk, plot_top_journals_data(ilk_bibliography)),
      tar_target(plot_top_journals_ilk, plot_top_journals(plot_top_journals_data_ilk)),
      tar_target(plot_top_country_data_ilk, plot_top_country_data(ilk_bibliography)),
      tar_target(plot_top_country_ilk, plot_top_country(plot_top_country_data_ilk)),
      tar_quarto(
        report,
        "bibliography_report.qmd",
        execute_params = list(),
        quiet = FALSE
      )
    )
   
Now I want to run this pipeline for several input files (initially 4) and new ones will be added over time.

If I understand branching correctly (https://books.ropensci.org/targets/dynamic.html) dynamic branching with the `map()` pattern is the way to go. But I am at a loss how I can implement this? 

To complicate things, the quarto document depends on the output from all previous targets mentioned, but not the ones based on other files, so it has e.g. `tar_read(plot_oa_status_ilk)` in it. 

I am at a loss here - how can I achieve this without repeating the same code block again and again?

I was pointed out to me by Eric Scott that the function `tar_files()` from the package `tarchetypes` would be the first stop, and now I have:

```
tar_files(
        files,
        c(
            file.path(".", "input", "IPBES ILK.csv"),
            file.path(".", "input", "IPBES IAS.csv")
        )
    ),
    tar_target(bibliography, load_bibliography(files), pattern = map(files))
```

which works and two targets `bibliography_xxxxxxxxxx` are created. Perfect. But how do I continue?

If I add 

```
tar_target(figure_top_journ_data, plot_top_journals_data(bibliography), pattern = map(files)) 
```

I get an error 

```
✔ skipped pattern bibliography
▶ dispatched branch figure_top_journ_data_3bf82b8344c09cb1
✖ errored branch figure_top_journ_data_3bf82b8344c09cb1
✖ errored pipeline [3.163 seconds]
Error:
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    no applicable method for 'group_by' applied to an object of class "NULL"
Last error traceback:
    plot_top_journals_data(bibliography)
    dplyr::slice_max(dplyr::filter(dplyr::rename(dplyr::summarise(dplyr::gro...
    dplyr::filter(dplyr::rename(dplyr::summarise(dplyr::group_by(bibliograph...
    dplyr::rename(dplyr::summarise(dplyr::group_by(bibliography$bibliography...
    dplyr::summarise(dplyr::group_by(bibliography$bibliography, Publication....
    dplyr::group_by(bibliography$bibliography, Publication.Title)
    .handleSimpleError(function (condition)  {     state$error <- build_mess...
    h(simpleError(msg, call))
```

but running 

```
plot_top_journals_data(tar_read(bibliography_3bf82b8344c09cb1))
```

works as expected?

**Comment by ginolhac** [comment, marked as answer]
Hello,
the map pattern should be following the correct targets. So this part

``` r
tar_target(figure_top_journ_data, plot_top_journals_data(bibliography), pattern = map(files)) 
```

should be 

``` r
tar_target(figure_top_journ_data, plot_top_journals_data(bibliography), pattern = map(bibliography)) 
```

**Reply by rkrug** [reply]
Thanks a lot - that makes perfect sense and it works.

Quick follow up: is there a way of getting the part of the name after the `_` for each file, ore change the naming pattern to be easier to understand? Follow up processing is in this case not to combine them, but to work with them separately.

**Reply by ginolhac** [reply]
yes, that is **Static** branching. https://books.ropensci.org/targets/static.html
If you have not so many things to iterate over it is nice to get clear names. But it is a bit more difficult to grasp at first.
I have some example with the `datasauRus` dataset here if it helps https://ginolhac.github.io/porg-targets/#/static-branching-with-dynamic-inside
